src/ResourceManager/DataLakeStore/Commands.DataLakeStore/Commands/ImportAzureRmDataLakeStoreItem.cs

```diff
@@ -82,49 +85,113 @@ public int ConcurrentFileCount
+        {
+            get { return logLevel; }
+            set { logLevel = value; }
+        }
+
+        [Parameter(ValueFromPipelineByPropertyName = true, Mandatory = false,
+            HelpMessage = "Optionally specifies the path for the diagnostic log to record events to during the file or folder import. If logging is enabled, the default is in %LOCALAPPDATA%\\ADLDataTransfer.")]
+        [ValidateNotNullOrEmpty]
+        public string DiagnosticLogPath { get; set; }
+
         public override void ExecuteCmdlet()
         {
             var powerShellSourcePath = SessionState.Path.GetUnresolvedProviderPathFromPSPath(Path);
+            string diagnosticPath = string.Empty;
+            if (!string.IsNullOrEmpty(DiagnosticLogPath))
+            {
+                diagnosticPath = SessionState.Path.GetUnresolvedProviderPathFromPSPath(DiagnosticLogPath);
+            }
+
+            // Setting -Debug overwrites LogLevel to debug.
+            if (MyInvocation.BoundParameters.ContainsKey("Debug") && (SwitchParameter)MyInvocation.BoundParameters["Debug"])
+            {
+                // TODO: Decide if we don't support outputting to the debug stream as well (since it is synchronous).
+                DiagnosticLogLevel = LogLevel.Debug;
+            }
+
             ConfirmAction(
                 Resources.UploadFileMessage,
                 Destination.TransformedPath,
                 () =>
                 {
-                    if (Directory.Exists(powerShellSourcePath))
+                    DataLakeStoreTraceLogger logger = null;
+                    var originalLevel = AdalTrace.TraceSource.Switch.Level;
+                    var originalLegacyLevel = AdalTrace.LegacyTraceSwitch.Level;
+                    try
                     {
-                        DataLakeStoreFileSystemClient.CopyDirectory(
-                            Destination.TransformedPath,
-                            Account,
-                            powerShellSourcePath,
-                            CmdletCancellationToken,
-                            ConcurrentFileCount,
-                            PerFileThreadCount,
-                            Recurse,
-                            Force,
-                            Resume, ForceBinary, ForceBinary, cmdletRunningRequest: this);
-                    }
-                    else if (File.Exists(powerShellSourcePath))
-                    {
-                        DataLakeStoreFileSystemClient.CopyFile(
-                            Destination.TransformedPath,
-                            Account,
-                            powerShellSourcePath,
-                            CmdletCancellationToken,
-                            PerFileThreadCount,
-                            Force,
-                            Resume,
-                            ForceBinary,
-                            cmdletRunningRequest: this);
+                        if (DiagnosticLogLevel != LogLevel.None)
```

Comment:
Agreed. I have moved everything into the constructor and dispose() methods of the logger for now. I think in the future (based on your suggestions) this can be refactored further.